clippy_lints/src/manual_ignore_case_cmp.rs

```diff
@@ -0,0 +1,98 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet;
+use clippy_utils::ty::{is_type_diagnostic_item, is_type_lang_item};
+use rustc_errors::Applicability;
+use rustc_hir::ExprKind::{Binary, MethodCall};
+use rustc_hir::{BinOpKind, Expr, LangItem};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty;
+use rustc_middle::ty::{Ty, UintTy};
+use rustc_session::declare_lint_pass;
+use rustc_span::sym;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for manual case-insensitive ASCII comparison.
+    ///
+    /// ### Why is this bad?
+    /// The `eq_ignore_ascii_case` method is faster because it does not allocate
+    /// memory for the new strings, and it is more readable.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// fn compare(a: &str, b: &str) -> bool {
+    ///     a.to_ascii_lowercase() == b.to_ascii_lowercase()
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// fn compare(a: &str, b: &str) -> bool {
+    ///    a.eq_ignore_ascii_case(b)
+    /// }
+    /// ```
+    #[clippy::version = "1.82.0"]
+    pub MANUAL_IGNORE_CASE_CMP,
+    nursery,
+    "manual case-insensitive ASCII comparison"
+}
+
+declare_lint_pass!(ManualIgnoreCaseCmp => [MANUAL_IGNORE_CASE_CMP]);
+
+fn get_ascii_type<'tcx>(cx: &LateContext<'tcx>, expr: &Expr<'_>) -> Option<Ty<'tcx>> {
+    let ty_raw = cx.typeck_results().expr_ty(expr);
+    let ty = ty_raw.peel_refs();
+    if needs_ref_to_cmp(cx, ty)
```

Comment:
Can you explain these lines?

They feel a bit weird. The `peel_refs` removes all potential references from the type, which means that some *ascii* types will fail the `needs_ref_to_cmp` function even if `ty_raw` would pass it.

Reply:
@xFrednet what's the better way to do this comparison?

Reply:
i was thinking some more on this, and tried to come up with a unit test that would fail, but was not able to do that... are you certain this is an issue?

Reply:
I think the naming tripped me up. When using this function you should make sure that the raw type is still a reference. `(ty_raw.is_ref() && needs_ref_to_cmp(cx, ty))` should do the trick

Reply:
@xFrednet I tried your suggestion, but it starts to miss half of the test cases like these:

```rust
// a & b are &str
a.to_uppercase().to_ascii_lowercase() == b.to_ascii_lowercase();